File: windows/Core/Network/DevControlSession.cs
using System.Buffers.Binary;
using System.Net;
using System.Net.Sockets;
using System.Text.Json;
using BarelyReal.Core.Protocol;

namespace BarelyReal.Core.Network;

public sealed class DevControlSession : IDisposable
{
    private TcpListener? _listener;
    private CancellationTokenSource? _cts;
    private string? _peerHost;
    private ushort _peerPort = 24800;

    public event Action<ScreenAnnouncement>? ScreenAnnounced;
    public event Action<LayoutSyncMessage>? LayoutSynced;
    public event Action<string>? LogLine;

    public void Start(ushort localPort = 24800, string? peerHost = null, ushort peerPort = 24800)
    {
        Close();
        _peerHost = peerHost?.Trim();
        _peerPort = peerPort;
        _cts = new CancellationTokenSource();
        _listener = new TcpListener(IPAddress.Any, localPort);
        _listener.Start();
        _ = AcceptLoop(_listener, _cts.Token);
        LogLine?.Invoke($"Control listening on TCP :{localPort}");
    }

    public void SendHello(HelloMessage hello) => SendJson(ControlType.Hello, hello);
    public void SendScreenAnnounce(ScreenAnnouncement announcement) => SendJson(ControlType.ScreenAnnounce, announcement);
    public void SendLayoutSync(LayoutSyncMessage layout) => SendJson(ControlType.LayoutSync, layout);
    public void SendKeepAlive() => Send(new ControlFrame(ControlType.KeepAlive, "{}"u8.ToArray()));

    public void Close()
    {
        _cts?.Cancel();
        _listener?.Stop();
        _cts?.Dispose();
        _cts = null;
        _listener = null;
    }

    public void Dispose() => Close();

    private async Task AcceptLoop(TcpListener listener, CancellationToken token)
    {
        try
        {
            while (!token.IsCancellationRequested)
            {
                var client = await listener.AcceptTcpClientAsync(token).ConfigureAwait(false);
                _ = Task.Run(() => HandleClient(client, token), token);
            }
        }
        catch (OperationCanceledException) { }
        catch (ObjectDisposedException) { }
        catch (Exception ex)
        {
            LogLine?.Invoke($"Control accept failed: {ex.Message}");
        }
    }

    private async Task HandleClient(TcpClient client, CancellationToken token)
    {
        using (client)
        using (var stream = client.GetStream())
        {
            try
            {
                var header = new byte[4];
                await stream.ReadExactlyAsync(header, token).ConfigureAwait(false);
                var length = BinaryPrimitives.ReadUInt32LittleEndian(header);
                if (length < 1 || length > ControlFrameCodec.MaxBodySize + 1)
                {
                    LogLine?.Invoke($"Control rejected invalid length {length}");
                    return;
                }
                var body = new byte[(int)length];
                await stream.ReadExactlyAsync(body, token).ConfigureAwait(false);
                var packet = new byte[4 + body.Length];
                BinaryPrimitives.WriteUInt32LittleEndian(packet.AsSpan(0, 4), length);
                body.CopyTo(packet.AsSpan(4));
                var (frame, _) = ControlFrameCodec.Decode(packet);
                Handle(frame);
            }
            catch (OperationCanceledException) { }
            catch (Exception ex)
            {
                LogLine?.Invoke($"Control receive failed: {ex.Message}");
            }
        }
    }

    private void Handle(ControlFrame frame)
    {
        try
        {
            switch (frame.Type)
            {
                case ControlType.ScreenAnnounce:
                    ScreenAnnounced?.Invoke(JsonSerializer.Deserialize<ScreenAnnouncement>(frame.Body)!);
                    break;
                case ControlType.LayoutSync:
                    LayoutSynced?.Invoke(JsonSerializer.Deserialize<LayoutSyncMessage>(frame.Body)!);
                    break;
                case ControlType.Hello:
                    var hello = JsonSerializer.Deserialize<HelloMessage>(frame.Body)!;
                    var peerId = hello.Os.Contains("win", StringComparison.OrdinalIgnoreCase) ? "windows" : "mac";
                    ScreenAnnounced?.Invoke(new ScreenAnnouncement(peerId, hello.Screens));
                    break;
                case ControlType.KeepAlive:
                    break;
                default:
                    LogLine?.Invoke($"Control ignored {frame.Type}");
                    break;
            }
        }
        catch (Exception ex)
        {
            LogLine?.Invoke($"Control parse failed: {ex.Message}");
        }
    }

    private void SendJson<T>(ControlType type, T value)
    {
        try
        {
            Send(new ControlFrame(type, JsonSerializer.SerializeToUtf8Bytes(value)));
        }
        catch (Exception ex)
        {
            LogLine?.Invoke($"Control encode failed: {ex.Message}");
        }
    }

    private void Send(ControlFrame frame)
    {
        var host = _peerHost;
        if (string.IsNullOrWhiteSpace(host)) return;
        _ = Task.Run(async () =>
        {
            try
            {
                using var client = new TcpClient();
                await client.ConnectAsync(host, _peerPort).ConfigureAwait(false);
                var payload = ControlFrameCodec.Encode(frame);
                await client.GetStream().WriteAsync(payload).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                LogLine?.Invoke($"Control send failed: {ex.Message}");
            }
        });
    }
}

File: mac/Core/Network/DevControlSession.swift
import Foundation
import Network

public final class DevControlSession: @unchecked Sendable {
    public enum ControlError: Error, Equatable {
        case invalidPort(UInt16)
    }

    public var onScreenAnnounce: ((ScreenAnnouncement) -> Void)?
    public var onLayoutSync: ((LayoutSyncMessage) -> Void)?
    public var onLog: ((String) -> Void)?

    private let queue: DispatchQueue
    private var listener: NWListener?
    private var peerHost: String?
    private var peerPort: UInt16

    public init(queue: DispatchQueue = DispatchQueue(label: "com.barelyreal.dev-control")) {
        self.queue = queue
        self.peerPort = 24_800
    }

    public func start(localPort: UInt16 = 24_800, peerHost: String?, peerPort: UInt16 = 24_800) throws {
        close()
        self.peerHost = peerHost?.trimmingCharacters(in: .whitespacesAndNewlines)
        self.peerPort = peerPort

        guard let port = NWEndpoint.Port(rawValue: localPort) else {
            throw ControlError.invalidPort(localPort)
        }

        let listener = try NWListener(using: .tcp, on: port)
        listener.newConnectionHandler = { [weak self] connection in
            self?.accept(connection)
        }
        listener.start(queue: queue)
        self.listener = listener
        onLog?("Control listening on TCP :\(localPort)")
    }

    public func close() {
        listener?.cancel()
        listener = nil
    }

    public func sendHello(_ hello: HelloMessage) {
        sendJSON(.hello, hello)
    }

    public func sendScreenAnnounce(_ announcement: ScreenAnnouncement) {
        sendJSON(.screenAnnounce, announcement)
    }

    public func sendLayoutSync(_ layout: LayoutSyncMessage) {
        sendJSON(.layoutSync, layout)
    }

    public func sendKeepAlive() {
        send(.init(type: .keepAlive, body: Data("{}".utf8)))
    }

    private func accept(_ connection: NWConnection) {
        connection.start(queue: queue)
        receiveHeader(on: connection)
    }

    private func receiveHeader(on connection: NWConnection) {
        connection.receive(minimumIncompleteLength: 4, maximumLength: 4) { [weak self] data, _, _, error in
            guard let self else { return }
            guard error == nil, let data, data.count == 4 else {
                connection.cancel()
                return
            }
            let length: UInt32 = data.readLE(at: 0)
            guard length >= 1, length <= ControlFrameCodec.maxBodySize + 1 else {
                self.onLog?("Control rejected invalid length \(length)")
                connection.cancel()
                return
            }
            self.receiveBody(length: Int(length), on: connection)
        }
    }

    private func receiveBody(length: Int, on connection: NWConnection) {
        connection.receive(minimumIncompleteLength: length, maximumLength: length) { [weak self] body, _, _, _ in
            defer { connection.cancel() }
            guard let self, let body, body.count == length else { return }
            var packet = Data(capacity: 4 + body.count)
            packet.appendLE(UInt32(body.count))
            packet.append(body)
            do {
                let decoded = try ControlFrameCodec.decode(packet).frame
                self.handle(decoded)
            } catch {
                self.onLog?("Control decode failed: \(error)")
            }
        }
    }

    private func handle(_ frame: ControlFrame) {
        do {
            switch frame.type {
            case .screenAnnounce:
                onScreenAnnounce?(try JSONDecoder().decode(ScreenAnnouncement.self, from: frame.body))
            case .layoutSync:
                onLayoutSync?(try JSONDecoder().decode(LayoutSyncMessage.self, from: frame.body))
            case .hello:
                let hello = try JSONDecoder().decode(HelloMessage.self, from: frame.body)
                let announcement = ScreenAnnouncement(peerId: hello.os.lowercased().contains("win") ? "windows" : "mac", screens: hello.screens)
                onScreenAnnounce?(announcement)
            case .keepAlive:
                break
            default:
                onLog?("Control ignored \(frame.type)")
            }
        } catch {
            onLog?("Control message parse failed: \(error)")
        }
    }

    private func sendJSON<T: Encodable>(_ type: ControlType, _ value: T) {
        do {
            let data = try JSONEncoder().encode(value)
            send(.init(type: type, body: data))
        } catch {
            onLog?("Control encode failed: \(error)")
        }
    }

    private func send(_ frame: ControlFrame) {
        guard let peerHost, !peerHost.isEmpty,
              let port = NWEndpoint.Port(rawValue: peerPort)
        else { return }
        let payload = ControlFrameCodec.encode(frame)
        let connection = NWConnection(to: .hostPort(host: .name(peerHost, nil), port: port), using: .tcp)
        connection.start(queue: queue)
        connection.send(content: payload, completion: .contentProcessed { [weak self] error in
            if let error {
                self?.onLog?("Control send failed: \(error)")
            }
            connection.cancel()
        })
    }
}
